### ISSUE TITLE
Caching on disk results in error.

### ISSUE THREAD

**Issue by Sumered** [ultralytics/ultralytics#8627]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi! 
First of all thanks for amazing work and effort that goes into this repository!

I have encountered a rather puzzling problem.
When using this training command:
```
model = YOLO("yolov8x.pt")
results = model.train(data = "/home/jovyan/work/ultralytics/DatasetYolo.yaml", epochs = 200, patience = 10, 
                      batch = 16, imgsz = 1280, save = True, save_period = 10, fraction=0.1,
                      cache = "disk", device = [0,1,2,3], name = "2024-03-02-Baseline")
```
It results in such error:
             
```
File "/home/jovyan/work/ultralytics/ultralytics/data/dataset.py", line 361, in save_dataset_cache_file
    path.with_suffix(".cache.npy").rename(path)  # remove .npy suffix
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/pathlib.py", line 1175, in rename
    os.rename(self, target)
FileNotFoundError: [Errno 2] No such file or directory: '/home/jovyan/work/ultralytics/Dataset/DatasetTrain.cache.npy' -> '/home/jovyan/work/ultralytics/Dataset/DatasetTrain.cache'
```

When I changed my training command to:
```
model = YOLO("yolov8x.pt")
results = model.train(data = "/home/jovyan/work/ultralytics/DatasetYolo.yaml", epochs = 200, patience = 10, 
                      batch = 16, imgsz = 1280, save = True, save_period = 10, fraction=0.1,
                      cache = "disk", device = [0,1], name = "2024-03-02-Baseline")
```

The error does not occur.

Now imporant part is that I edited the code of base.py (line 145, function get_img_files) in data like this:
```
if self.fraction < 1:
            im_files = random.sample(im_files, round(len(im_files) * self.fraction))
```
The reasoning is mentioned here: [Fraction parameter](https://github.com/ultralytics/ultralytics/issues/6464)
So I changed it back to original code which is:
```
if self.fraction < 1:
            im_files = im_files[: round(len(im_files) * self.fraction)]
```
It worked correctly.

So for the sake of checking different options I replaced this with:
```
if self.fraction < 1:
            random.Random(1007).shuffle(im_files)
            im_files = im_files[: round(len(im_files) * self.fraction)]
```
And now it works correctly (caching does not throw errors) and training proceeds without errors.

What is also interesting is that I already trained dozens of models using this change to code and it never resulted in error. What changed is that now I'm using 4 gpus instead of 2.

I edited this code with some intuitions but I'm not really sure why this change results in no error? Obviously this bug is the result of my code change, but if You could provide me with information how does data loading exactly happens in ultralytics this could possibly shed some light on this.

Thanks!

**Comment by glenn-jocher** [comment]
@Sumered hi there! 🌟

Thank you for the kind words and for diving deep into the YOLOv8 codebase! It's great to see such engagement.

Regarding your issue, it seems like the error with caching on disk arises due to the way the dataset is being sampled and cached across multiple GPUs. When you use `random.sample()`, it likely introduces non-determinism in the order of image files across different processes, leading to inconsistencies when trying to cache or access the cache file. This non-determinism is exacerbated when the number of GPUs is changed.

Your solution to shuffle the image files with a fixed seed (`random.Random(1007).shuffle(im_files)`) before slicing ensures that all processes see the image files in the same order, thus avoiding the caching error. This approach maintains determinism across multiple runs and GPU configurations.

Data loading in YOLOv8, especially with caching enabled, relies on consistent access to the dataset across all worker processes. Any change in the dataset order or availability can lead to errors when trying to read from or write to the cache file.

In summary, your intuitive fix aligns with best practices for parallel data processing: ensuring consistency and determinism across all processes. If you're experimenting with modifications to the data loading logic, keeping these principles in mind will help avoid similar issues.

Keep up the great work, and don't hesitate to reach out if you have more questions! 🚀
